src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1029 @@
+            event_queue.put(change)
+
+    def send_hearbeat_msg(self):
+        """Send a health heartbeat event"""
+        if self.num_osds > 0 and self.raw_capacity == 0:
+            # we have osds, but the stats aren't available yet, so skip the heartbeat
+            return
+
+        health_str = self.mgr.get('health')['json']
+        if json.loads(health_str)['status'] == 'HEALTH_OK':
+            health_text = 'Healthy'
+            level = 'INF'
+        else:
+            health_text = 'Unhealthy'
+            level = 'WRN'
+
+        health_msg = ("Cluster state: {}. {} host{}, {} pool{}, {} OSDs. Raw "
+                      "Capacity {}B".format(health_text, self.num_servers, text_suffix(self.num_servers),
+                                            self.num_pools, text_suffix(self.num_pools),
+                                            self.num_osds, MgrModule.to_pretty_iec(self.raw_capacity)))
+
+        event_queue.put(
+            LogEntry(
+                source="config",
+                msg_type="heartbeat",
+                msg=health_msg,
+                level=level,
+                tstamp=None
+            )
+        )
+
+    def _generate_logentry(self, msg):
+        return LogEntry(
+                source="config",
+                msg_type="config",
+                msg=msg,
+                level='INF',
+                tstamp=None
+        )
+    
+    def _check_hosts(self, server_map):
+        changes = list()
+        if set(self.server_map.keys()) == set(server_map.keys()):
+            # no hosts changes
+            pass
+        else:
+            # host changes detected, find out what
+            diff = ListDiff(self.server_map.keys(), server_map.keys())
+            host_msg = "Host '{}' has been {} the cluster"
+            for new in diff.added:
+                changes.append(self._generate_logentry(
+                                    msg=host_msg.format(new, 'added to'))
+                )
+
+            for removed in diff.removed:
+                changes.append(self._generate_logentry(
+                                    msg=host_msg.format(removed, 'removed from'))
+                )
+
+        return changes
+
+    def _check_osds(self,server_map, osd_map):
+        changes = list()
+        before_osds = list()
+        for svr in self.server_map:
+            before_osds.extend(self.server_map[svr].get('osd',[]))
+
+        after_osds = list()
+        for svr in server_map:
+            after_osds.extend(server_map[svr].get('osd',[]))
+        
+        if set(before_osds) == set(after_osds):
+            # no change in osd id's
+            pass
+        else:
+            # osd changes detected
+            osd_msg = "Ceph OSD '{}' ({} @ {}B) has been {} host {}"
+
+            diff = ListDiff(before_osds, after_osds)
+            for new in diff.added:
+                changes.append(self._generate_logentry(
+                                    msg=osd_msg.format(
```

Comment:
This 
```suggestion
                                    msg="Ceph OSD '{}' ({} @ {}B) has been {} host {}".format(
```
would be an idea to allow your fellow compiler to check that the number of arguments mach the number of `{}`s

Reply:
? I put the format into a string since it's used multiple times with only slight variance to cut down of changes